assembly/src/sema/mod.rs

```diff
@@ -0,0 +1,206 @@
+                    analyzer.error(SemanticAnalysisError::UnusedDocstring {
+                        span: unused.span(),
+                    });
+                }
+            }
+            Form::Constant(constant) => {
+                analyzer.define_constant(constant.with_docs(docs.take()))?;
+            }
+            Form::Import(import) => {
+                if let Some(docs) = docs.take() {
+                    analyzer.error(SemanticAnalysisError::ImportDocstring { span: docs.span() });
+                }
+                analyzer.define_import(import)?;
+            }
+            Form::Procedure(export @ Export::Alias(_)) => match kind {
+                ModuleKind::Kernel => {
+                    docs.take();
+                    analyzer.error(SemanticAnalysisError::ReexportFromKernel {
+                        span: export.span(),
+                    });
+                }
+                ModuleKind::Executable => {
+                    docs.take();
+                    analyzer.error(SemanticAnalysisError::UnexpectedExport {
+                        span: export.span(),
+                    });
+                }
+                ModuleKind::Library => {
+                    analyzer.define_procedure(export.with_docs(docs.take()))?;
+                }
+            },
+            Form::Procedure(export) => match kind {
+                ModuleKind::Executable
+                    if export.visibility().is_exported() && !export.is_main() =>
+                {
+                    docs.take();
+                    analyzer.error(SemanticAnalysisError::UnexpectedExport {
+                        span: export.span(),
+                    });
+                }
+                _ => {
+                    analyzer.define_procedure(export.with_docs(docs.take()))?;
+                }
+            },
+            Form::Begin(body) if matches!(kind, ModuleKind::Executable) => {
+                let docs = docs.take();
+                let source_file = analyzer.source_file();
+                let procedure =
+                    Procedure::new(body.span(), Visibility::Public, ProcedureName::main(), 0, body)
+                        .with_docs(docs)
+                        .with_source_file(Some(source_file));
+                analyzer.define_procedure(Export::Procedure(procedure))?;
+            }
+            Form::Begin(body) => {
+                docs.take();
+                analyzer.error(SemanticAnalysisError::UnexpectedEntrypoint { span: body.span() });
+            }
+        }
+    }
+
+    if let Some(unused) = docs.take() {
+        analyzer.error(SemanticAnalysisError::UnusedDocstring {
+            span: unused.span(),
+        });
+    }
+
+    if matches!(kind, ModuleKind::Executable) && !analyzer.module().has_entrypoint() {
+        analyzer.error(SemanticAnalysisError::MissingEntrypoint);
+    }
+
+    if analyzer.has_errors() {
+        return analyzer.into_result();
+    }
```

Comment:
Somewhat related to the previous comment:

The way `into_result()` is used here (and a few lines below) is a bit odd. Specifically, we call it only in contexts when we know it'll return errors. Maybe a cleaner approach would be to have `into_errors()` and then this code would become:
```Rust
if analyzer.has_errors() {
    return analyzer.into_errors();
}
```

Reply:
I think this can actually just become:

```
analyzer.into_result()?;
```

Since the check for `has_errors` here is redundant